lib/ember-app.js

```diff
@@ -54,30 +55,49 @@ function EmberApp(options) {
   this._app = AppFactory['default']();
 }
 
-EmberApp.prototype.boot = function() {
-  return this._app.boot();
+EmberApp.prototype.boot = function(options) {
+  return this._app.boot().then(function(app) {
+    return app.buildInstance().boot(options);
+  });
 };
 
-EmberApp.prototype.visit = function(url) {
+EmberApp.prototype.visit = function(request, response) {
+  var path = request.path;
   var doc = new SimpleDOM.Document();
   var rootElement = doc.body;
   var options = { isBrowser: false, document: doc, rootElement: rootElement };
 
-  return this._app.visit(url, options).then(function(instance) {
-    var head;
-    if (doc.head) {
-      head = HTMLSerializer.serialize(doc.head);
-    }
-    try {
-      return {
-        url: instance.getURL(), // TODO: use this to determine whether to 200 or redirect
-        title: doc.title,
-        head: head,
-        body: HTMLSerializer.serialize(rootElement) // This matches the current code; but we probably want `serializeChildren` here
-      };
-    } finally {
-      instance.destroy();
+  return this.boot(options).then(function(instance) {
+    var fastbootService = instance.lookup('service:fastboot');
+
+    var cookies = request.get('cookie');
+    if (cookies) {
+      cookies = cookie.parse(cookies);
     }
+
+    fastbootService.setProperties({
```

Comment:
I think restoring authentication state (or user settings etc.) from cookies in (instance) initializers is probably still a pretty common thing to do. In fact, Ember Simple Auth did just that until version 1.0 as well.

I think the `ApplicationInstance` returned from `buildInstance` (https://github.com/ember-fastboot/ember-fastboot-server/pull/19/files#diff-ba82abedace362f3becf198dd49dacb4R60) already has the `register` method so that it should be possible to register the request and response as `-request:main` and `-response:main` and inject these into the `fastboot` service then?